presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java

```diff
@@ -0,0 +1,1970 @@
+                    break;
+                case FINISHED:
+                    throw new IllegalStateException("Tried to commit buffered metastore operations after transaction has been committed/aborted");
+                default:
+                    throw new IllegalStateException("Unknown state");
+            }
+        }
+        finally {
+            state = State.FINISHED;
+        }
+    }
+
+    public synchronized void rollback()
+    {
+        try {
+            switch (state) {
+                case EMPTY:
+                case EXCLUSIVE_OPERATION_BUFFERED:
+                    break;
+                case SHARED_OPERATION_BUFFERED:
+                    rollbackShared();
+                    break;
+                case FINISHED:
+                    throw new IllegalStateException("Tried to rollback buffered metastore operations after transaction has been committed/aborted");
+                default:
+                    throw new IllegalStateException("Unknown state");
+            }
+        }
+        finally {
+            state = State.FINISHED;
+        }
+    }
+
+    private void commitShared()
+    {
+        // If exception is thrown late enough, it makes more sense to move forward than backward
+        boolean moveForwardInFinally = false;
+
+        AtomicBoolean fileRenameCancelled = new AtomicBoolean(false);
+        List<CompletableFuture<?>> fileRenameFutures = new ArrayList<>();
+
+        // File system
+        List<DirectoryCleanUpTask> cleanUpTasks = new ArrayList<>();
+        List<DirectoryRenameTask> renameTasks = new ArrayList<>();
+
+        // Metastore
+        List<CreateTableOperation> addTableOperations = new ArrayList<>();
+        List<PartitionAdder> partitionAdders = new ArrayList<>();
+        List<AlterPartitionOperation> alterPartitionOperations = new ArrayList<>();
+        List<IrreversibleMetastoreOperation> metastoreDeleteOperations = new ArrayList<>();
+
+        try {
+            for (Map.Entry<SchemaTableName, Action<TableAndMore>> entry : tableActions.entrySet()) {
+                SchemaTableName schemaTableName = entry.getKey();
+                Action<TableAndMore> action = entry.getValue();
+                switch (action.getType()) {
+                    case DROP:
+                        metastoreDeleteOperations.add(new IrreversibleMetastoreOperation(
+                                format("drop table %s.%s", schemaTableName.getSchemaName(), schemaTableName.getTableName()),
+                                () -> delegate.dropTable(schemaTableName.getSchemaName(), schemaTableName.getTableName(), true)));
+                        break;
+                    case ALTER:
+                        throw new UnsupportedOperationException("Dropping and then creating a table with the same name is not supported");
+                    case ADD: {
+                        Table table = action.getData().getTable();
```

Comment:
These three variables are only used in a couple of locations.  Consider just inlining them